tainting: Analyze lambdas in context

IagoAbal opened the pull request:
Lambdas were already analyzed in isolation, as they are function
definitions, but they were not being analyzed in the context surrounding
them. So if e.g. some variable was tainted outside a lambda, and reached
a sink inside the lambda, this was not being detected.

Helps PA-1224
Closes PA-1308
Closes #5239

test plan:
make test # added two tests

PR checklist:

- [ ] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

IagoAbal committed:
tainting: Analyze lambdas in context

Lambdas were already analyzed in isolation, as they are function
definitions, but they were not being analyzed in the context surrounding
them. So if e.g. some variable was tainted outside a lambda, and reached
a sink inside the lambda, this was not being detected.

Helps PA-1224
Closes PA-1308
Closes #5239

test plan:
make test # added two tests

IagoAbal committed:
Fix test expectations in tests/semgrep-rules

PR returntocorp/semgrep-rules#2108

IagoAbal commented:
[run semgrep on public repos (3.7)](https://github.com/returntocorp/semgrep/pull/5348#logs) fails due to a timeout error running a dummy pattern on DevSlop-Pixi-latest/app/public/api-docs/swagger-ui-standalone-preset.js. This is a minimized JS file. There is a function definition that contains 800+ lambdas, and the resulting graph has now 17,229 nodes and takes ~150s to run constant propagation on it. Problem seems to be an excess of allocations during the fixpoint computation, e.g. major collections of from 14 to 529! Promoted words go from ~20 M to ~3k M!

I presume this may be the same problem as with benchmark semgrep.bench.Vulnerable-Flask-App.std. I'm investigating how to optimize this. Otherwise we could skip const-prop for large CFGs.

aryx commented on CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 - Sarif output format now includes `fixes` section
 - Rust: added support for method chaining patterns.
+- taint-mode: Taint tracking will now analyze lambdas in their surrounding context.
+  Previously, if a variable became tainted outside a lanbda, and this variable was
```
typo, lanbda

aryx commented on CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
+  used inside the lambda causing the taint to reach a sink, this was not being
+  detected because any nested lambdas were "opaque" to the analysis. (Taint tracking
+  looked at lambdas but as isolated functions.) Now lambas are simply analyzed as if
+  they were statement blocks. Taint tracking still does not follow the flow of taint
```
However, taint tracking ... or though at the end of the sentence.

aryx commented:
The CI failures are about semgrep-rules, but as you updated the local semgrep-rules submodule, and those test pass, I guess we can merge this.

IagoAbal committed:
const-prop: Save allocations by not storing NotCst

test plan:
1. Check out https://github.com/DevSlop/Pixi
   ^ Tested with cb716a94fa368acade96e9bdb8b25c1a68703e60
2. semgrep-core semgrep -e '$SENTINEL = 87518275812375164' -lang javascript app/public/api-docs/swagger-ui-standalone-preset.js
  ^ This now takes ~2s instead of ~150s !!!!!

IagoAbal committed:
Yoann's review

IagoAbal commented:
OK then [semgrep unit and e2e tests (3.7)](https://github.com/returntocorp/semgrep/pull/5348#logs) only reports:

> FAILED tests/e2e/test_semgrep_rules_repo.py::test_semgrep_rules_repo - Failed...

and [semgrep tests with different python envs (3.9)](https://github.com/returntocorp/semgrep/pull/5348#logs) reports:

> E   	✖ unsafe-argon2-config                                         expected lines: [47], reported lines: [32, 47] 
[2204](https://github.com/returntocorp/semgrep/runs/6695406036?check_suite_focus=true#step:8:2205)
> E   	 test file path: /tmp/pytest-of-runner/pytest-0/popen-gw1/test_semgrep_rules_repo0/semgrep-rules/javascript/argon2/security/unsafe-argon2-config.js

This should be fixed in https://github.com/returntocorp/semgrep-rules/pull/2108 (which is already approved by SR).

So I'll merge!
